### PR TITLE
CD-112773 Add methods for verifying tokens with a specified resource …

### PR DESCRIPTION
**Pull request by johnny-lai**
…and context

- [x] @johnwu96822

### CODE REVIEW

File: sand.go [hunk 1]
```diff
@@ -122,8 +122,10 @@ func (c *Client) RequestWithCustomRetry(cacheKey string, scopes []string, numRet
 //Token returns an OAuth token retrieved from the OAuth2 server. It also puts the
 //token in the cache up to specified amount of time.
 func (c *Client) Token(cacheKey string, scopes []string, numRetry int) (string, error) {
+  var ckey string
 	if c.Cache != nil && cacheKey != "" {
-		token := c.Cache.Read(c.cacheKey(cacheKey, scopes))
+    ckey = c.cacheKey(cacheKey, scopes, "")
```
Comment (johnwu96822) [inline sand.go]: gofmt

File: sand.go [hunk 2]
```diff
@@ -183,10 +185,13 @@ func (c *Client) oauthToken(scopes []string, numRetry int) (token *oauth2.Token,
 }
 
 //cacheKey builds the cache key in the format: <CachRoot>/<cacheType>/<key>
-func (c *Client) cacheKey(key string, scopes []string) string {
+func (c *Client) cacheKey(key string, scopes []string, resource string) string {
 	rv := c.CacheRoot + "/" + c.cacheType + "/" + key
 	if len(scopes) > 0 {
 		rv += "/" + strings.Join(scopes, "_")
 	}
+	if len(resource) > 0 {
```
Comment (johnwu96822) [inline sand.go]: For string I typically do: `if resource != ""`

File: service.go [hunk 3]
```diff
@@ -119,16 +137,16 @@ func (s *Service) isTokenAllowed(token string, targetScopes []string, action str
...
-			s.Cache.Write(s.cacheKey(token, targetScopes), notAllowedResponse, time.Duration(s.DefaultExpTime)*time.Second)
+			s.Cache.Write(ckey, notAllowedResponse, time.Duration(s.DefaultExpTime)*time.Second)
 		}
 	}
 	return resp, nil
 }
 
 //verifyToken verifies with SAND to see if the token is allowed to access this service.
-func (s *Service) verifyToken(token string, targetScopes []string, action string, numRetry int) (map[string]interface{}, error) {
+func (s *Service) verifyTokenForResourceContext(token string, targetScopes []string, action string, resource string, context map[string]interface{}, numRetry int) (map[string]interface{}, error) {
```
Comment (johnwu96822) [inline service.go]: Return error if resource is empty string

File: sand.go [hunk 4]
```diff
@@ -183,10 +185,13 @@ func (c *Client) oauthToken(scopes []string, numRetry int) (token *oauth2.Token,
 }
 
 //cacheKey builds the cache key in the format: <CachRoot>/<cacheType>/<key>
-func (c *Client) cacheKey(key string, scopes []string) string {
+func (c *Client) cacheKey(key string, scopes []string, resource string) string {
 	rv := c.CacheRoot + "/" + c.cacheType + "/" + key
-	if len(scopes) > 0 {
+	if scopes != "" {
```
Comment (johnwu96822) [inline sand.go]: scopes is a string array

**Comment by johnwu96822** [comment]
👍 after last comment is resolved

File: sand.go (hunk 4 above)
Comment (johnny-lai) [reply sand.go]: Oops. my bad. thanks

**Comment by johnny-lai** [comment]
```
~/Coupa/go/src/github.com/coupa/sand-go (master_CD-112773)
johnny_lai$ go test
Running Suite: Sand Suite
=========================
Random Seed: 1526677517
Will run 44 of 44 specs

••WARN[0000] Sand request: retrying after 1 sec on 401
WARN[0001] Sand request: retrying after 2 sec on 401
••••WARN[0003] Sand request: retrying after 1 sec on 401
WARN[0004] Sand request: retrying after 1 sec on 401
WARN[0005] Sand request: retrying after 2 sec on 401
•••••••••WARN[0007] Sand token: retrying after 1 sec because of error: oauth2: cannot fetch token: 404 Not Found
Response:
WARN[0008] Sand token: retrying after 2 sec because of error: oauth2: cannot fetch token: 404 Not Found
Response:
•••••••ERRO[0010] auth error: Post : unsupported protocol scheme ""
•ERRO[0010] auth error: Post : unsupported protocol scheme ""
•WARN[0010] Sand token: retrying after 1 sec because of error: Post : unsupported protocol scheme ""
WARN[0011] Sand token: retrying after 2 sec because of error: Post : unsupported protocol scheme ""
ERRO[0013] auth error: Post : unsupported protocol scheme ""
••••••••••••••••••••
Ran 44 of 44 Specs in 13.046 seconds
SUCCESS! -- 44 Passed | 0 Failed | 0 Pending | 0 Skipped PASS
ok  	github.com/coupa/sand-go	13.426s
```

**Comment by johnny-lai** [comment]
[lockit v4.3.0] Merged via Lockit CLI on 2018-05-18 14:06:02.116412 -0700 PDT m=+5.030458935

**Pull request merged**